### PR TITLE
sql: fix limits on virtual index scans

### LINKED ISSUE #53522: sql: virtual tables don't respect LIMIT when there is a WHERE clause
**To reproduce**

It might relate to the indexes that were added to vtables.

```
root@:26257/defaultdb> create table t (id int primary key, name text, notselected text, ts timestamp with time zone, intarr int[]);
CREATE TABLE

Time: 74.401ms

root@:26257/defaultdb> select oid from pg_class where relname = 't';;
  oid
-------
  800
(1 row)

Time: 54.675ms

root@:26257/defaultdb> select attname from pg_attribute limit 1;
     attname
-----------------
  descriptor_id
(1 row)

Time: 52.461ms

root@:26257/defaultdb> select attname from pg_attribute where attrelid = 800 limit 1;
    attname
---------------
  id
  name
  notselected
  ts
  intarr
(5 rows)
```

On v20.2

### PR DESCRIPTION
**Pull request by jordanlewis**
Previously, virtual indexes mis-reported their key columns as containing
only their index key. Virtual indexes always need to contain the dummy
PK column since they're not guaranteed to be unique. This caused some
broken behavior like limits not working properly in some cases.

Fixes #53522 

Release note (bug fix): virtual index scans respect limits properly as
they did in 20.1 and prior.

Release justification: bug fix to new functionality

### CODE REVIEW

**Review by rafiss — CHANGES_REQUESTED** [review]
could you add a SQL logictest as well?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53529#-:-MFheN9AD637DV1cg_3O:bpr4l54)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @RaduBerinde)

___
*[pkg/storage/enginepb/mvcc3.pb.go, line 4279 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53529#-MFhdua19OSZtAGUYcOB:-MFhdua19OSZtAGUYcOC:bbhlna6) ([raw file](https://github.com/cockroachdb/cockroach/blob/85665f54318015353f65e38ac6fb896b50bdd40f/pkg/storage/enginepb/mvcc3.pb.go#L4279)):*
> ```Go
> 
> func init() {
> 	proto.RegisterFile("storage/enginepb/mvcc3.proto", fileDescriptor_mvcc3_ffb4a3e3dc0d81a7)
> ```

accidentally committed the pb.go files? (i think you can resolve this extra diff by doing `rm bin/bin/protoc-gen-gogoroach bin/gofmt bin/goimports` and re-compiling, and make sure you're on go1.13

**Review by jordanlewis — COMMENTED** [review]
Added logic test

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53529#-:-MFhiM5iCz_SU-cfKOXj:bnwtahl)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde and @rafiss)

___
*[pkg/storage/enginepb/mvcc3.pb.go, line 4279 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53529#-MFhdua19OSZtAGUYcOB:-MFhiJ0cDKUEGDFCXCva:b-q69455) ([raw file](https://github.com/cockroachdb/cockroach/blob/85665f54318015353f65e38ac6fb896b50bdd40f/pkg/storage/enginepb/mvcc3.pb.go#L4279)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

accidentally committed the pb.go files? (i think you can resolve this extra diff by doing `rm bin/bin/protoc-gen-gogoroach bin/gofmt bin/goimports` and re-compiling, and make sure you're on go1.13
</blockquote></details>

Oops, been meaning to fix this. Done.

File: pkg/sql/logictest/testdata/logic_test/pg_catalog [hunk 1]
```diff
@@ -2665,3 +2665,10 @@ WHERE
   AND attname = 'x'
 ----
 'howdy'::STRING
+
+# Regression test for limits on virtual index scans. (#53522)
+
+query T
+SELECT attname FROM pg_attribute WHERE attrelid = 't47285'::REGCLASS::OID LIMIT 1
```
Comment (rafiss) [inline pkg/sql/logictest/testdata/logic_test/pg_catalog:2672]: oddly enough, it looks like this query is _not_ affected by the original issue. it seems like the casting makes the query avoid hitting the problem. is there a way to write a test without casting? idk how, so if not, the tests lgtm

example from me testing locally with v20.2.0-alpha:
```
root@:26257/defaultdb> SELECT attrelid, attname FROM pg_attribute WHERE attrelid = 't'::REGCLASS::OID limit 1;
  attrelid | attname
-----------+----------
       215 | name
(1 row)

Time: 66.961ms

root@:26257/defaultdb> SELECT attrelid, attname FROM pg_attribute WHERE attrelid = 215 limit 1;
  attrelid |    attname
-----------+-----------------
       215 | name
       215 | username
       215 | security_level
       215 | last_login
       215 | locked
       215 | rowid
(6 rows)
```

**Review by rafiss — COMMENTED** [review]
had a comment on tests -- i'm not reviewing the code changes, but nothing blocking from my side

**Comment by jordanlewis** [comment]
Good point, I guess it turned into a normal filter with that. Rejiggered.

File: pkg/sql/logictest/testdata/logic_test/pg_catalog [hunk 2]
```diff
@@ -2665,3 +2665,13 @@ WHERE
   AND attname = 'x'
 ----
 'howdy'::STRING
+
+# Regression test for limits on virtual index scans. (#53522)
+
+let $testid
```
Comment (rafiss) [inline pkg/sql/logictest/testdata/logic_test/pg_catalog:2671]: glad i learned this syntax!

**Review by rafiss — APPROVED** [review]

**Review by mgartner — APPROVED** [review]
:lgtm: 

> Reviewed 1 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53529#-:-MFm8cCr6EE--3nKlX9e:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis, @RaduBerinde, and @rafiss)

___
*[pkg/sql/opt_catalog.go, line 1809 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/53529#-MFm6y8zCW92lads1XD5:-MFm6y8zCW92lads1XD6:bgqp16y) ([raw file](https://github.com/cockroachdb/cockroach/blob/71fae5ecd8588f075fbb6e359b731ba900ffa062/pkg/sql/opt_catalog.go#L1809)):*
> ```Go
> // KeyColumnCount is part of the cat.Index interface.
> func (oi *optVirtualIndex) KeyColumnCount() int {
> 	return 2
> ```

[nit] explain why a virtual index always has two key columns (and lax key columns below).
___
*[pkg/sql/opt_catalog.go, line 1839 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/53529#-MFm8RwE6Gj0J3-7wpuC:-MFm8RwE6Gj0J3-7wpuD:b-mlulus) ([raw file](https://github.com/cockroachdb/cockroach/blob/71fae5ecd8588f075fbb6e359b731ba900ffa062/pkg/sql/opt_catalog.go#L1839)):*
> ```Go
> 			Column: oi.tab.Column(ord),
> 		}
> 	} else if i == length {
> ```

[nit] the `else if` could just be an `if` since the previous `if` returns early.

**Review by jordanlewis — COMMENTED** [review]
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53529#-:-MG4eMbR6YTe8Sn7cM5f:b-oduykh)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis, @RaduBerinde, and @rafiss)

**Pull request merged**